api-proxy/src/server.mjs

```diff
@@ -0,0 +1,101 @@
+import * as http from 'http'
+import createLogger from './logging.mjs'
+import {handleReq} from './reqhandler.mjs'
+
+const nl = '\n'
+
+const parseUrl = (protocol, req) => {
+  const {host} = req.headers
+  if (!host) {
+    throw Err('no-host', 'missing Host header', {status: 400})
+  }
+  if (host.startsWith(protocol+'://')) {
+    return new URL(req.url, host)
+  } else {
+    return new URL(req.url, protocol+'://'+host)
+  }
+}
+
+
+export const createReqHandler = handleReq => {
+  let requestIndex = 0
+  const {log} = createLogger()
+  return async (req, res) => {
+    req.startTimeMs = Date.now()
+    req.index = requestIndex++
+    res.index = req.index
+    req.log = (...formatArgs) => {
+      const [f, ...args] = formatArgs
+      log('#%d '+f, req.index, ...args)
+    }
+    res.log = req.log
+    const protocol = req.socket.encrypted ? 'https' : 'http'
+    const {host} = req.headers
+    const isIp = host.match(/^\d+[.]\d+[.]\d+[.]\d+/)
+    if (isIp) { return res.status(400).typeText().mwrite('Invalid host.'+nl).mend() }
```

Comment:
why is IP invalid here?

Reply:
This only happens when your server is doing the host routing and is therefore exposed to all of the weird crap coming through the internet. It's not going to happen for us, so I removed the check.